### DISCUSSION TITLE
zero-install / cloud-IDE option for getting started with agoric-sdk

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#7033] [Ideas]
Goal: let devs try out smart contract development using only a web browser, without installing any dev tools.

a la [jsfiddle](https://jsfiddle.net/), [rust playground](https://play.rust-lang.org/) etc.

----

I just got it working on gitpod. Details to follow.

https://github.com/dckc/agoric-gitpod-nix

cc @michaelfig @katelynsills @rowgraus

**Comment by katelynsills** [comment]
Cool! What would this mean, exactly? Would it be for smart contract development or dapp development?

**Comment by dckc** [comment]
Either one, IIUC.

It means someone could follow a link to a web site that acts just like vs-code, with agorc-sdk, yarn, node, and go already installed and ready so that `yarn test` works. `agoric init demo` should work too.

I think I tested `agoric start` with the wallet and fake chain and all that.

**Comment by katelynsills** [comment]
Ooo that sounds great!

**Comment by dckc** [comment]
connecting to the wallet doesn't seem to quite work...

```
2021-09-16T02:28:02.880Z web: 127.0.0.1:37292: Invalid origin host https://8000-purple-clam-arf9beuf.ws-us15.gitpod.io is not localhost
```

**Comment by dckc** [comment]
Here's a 20 minute live-coding session that shows me getting as far as the wallet connecting problem above:

 - [agoric-sdk on Gitpod: zero-install BLD tools (WIP)](https://www.twitch.tv/videos/1150004318)

_ew... the video quality is horrible. I must have told OBS to use low resolution or something._

**Comment by dckc** [comment]
Yes! #3838 gets us past the localhost issue:

![image](https://user-images.githubusercontent.com/150986/133692749-f64df6e2-718a-48c8-a844-d1c6187d2634.png)

**Comment by dckc** [comment]
I'm not the only one having this funky terminal font issue. This [settings tweak](https://github.com/gitpod-io/gitpod/issues/5014#issuecomment-919901497) works for me just now:

```json
"terminal.integrated.gpuAcceleration": "canvas"
```

**Comment by dckc** [comment]
@michaelfig with the wallet running in some gitpod origin, how do I tell a dapp such as the fungible faucet demo to talk to it?

**Comment by michaelfig** [comment]
That's a yet-to-be-solved problem for browsers like Brave that aggressively block iframe service workers, cookies, and local storage, but with Chrome or Firefox, you can navigate your browser to https://local.agoric.com and fill out the form to point to your wallet with gitpod origin.

**Comment by samsiegart** [comment]
The wallet connection works for me with the local.agoric.com method above, but the faucet doesn't connect to the API server:
```
WebSocket connection to 'ws://127.0.0.1:8000/api/fungible-faucet' failed
```
We need to somehow dynamically change the hardcoded origin to use the gitpod one.

**Comment by dckc** [comment]
One of the big challenges in this area is that we don't support the `npm install -g agoric` convention because things move too quickly for that to be cost-effective: it would only install the version most recently published to npm, which is rarely what we (the Agoric dev team) want. This is a known issue: https://github.com/Agoric/agoric-sdk/issues/3857

**Comment by dckc** [comment]
another option to look into: https://stackblitz.com/

**Comment by dckc** [comment]
related work:
https://mobile.twitter.com/adamleetucker17/status/1513631602269724675
https://marketplace.digitalocean.com/apps/osmosis

**Comment by Tyrosine22** [comment]
I'm confused. Why is this item in the Doc board? It doesn't seem doc-related.

**Comment by dckc** [comment]
It's related to getting started. As you're discovering, I have used this issues list for all sorts of devrel stuff. Seems like it's time to be a little more consistent about scope. This sort of stuff can/should move to an agoric-sdk discussion. (I'm on a small screen just now but I can get to it in due course.)

**Comment by dckc** [comment]
@danBamikiya points out that [the Agoric vs-code extension works on gitpod](https://github.com/Agoric/documentation/pull/775#discussion_r1109063317).

I just tried it... sure enough, the REPL works:

![image](https://user-images.githubusercontent.com/150986/219913859-a9560f2d-825c-4353-8304-517ee799b8d2.png)

## URL surgery for `agoric open`

It took a bit of URL surgery to map from localhost...

```
gitpod ~/demo1 (master) $ agoric open
Launching wallet...
http://127.0.0.1:8000/wallet/#accessToken=qNU8kb5-kgwhmxo2bCACSTF8xymPr3Z77q0zJsyIy-eWTyveMlcgZhn-B67sDaHm
gitpod ~/demo1 (master) $ # https://8000-dckc-ibclab-db8zm275ytj.ws-us87.gitpod.io/#accessToken=qNU8kb5-kgwhmxo2bCACSTF8xymPr3Z77q0zJsyIy-eWTyveMlcgZhn-B67sDaHm
```

## Wallet config?

I bailed on trying to configure the wallet. Symptoms are similar to...

 - https://github.com/Agoric/agoric-sdk/issues/6189

See also:

https://github.com/Agoric/agoric-sdk/blob/master/packages/wallet/README.md
https://github.com/Agoric/agoric-sdk/pull/6498

**Comment by dckc** [comment]
## Hardened JavaScript works on Glitch

https://glitch.com/edit/#!/rainy-voracious-fiber

p.s. the more relevant issue is
 - https://github.com/Agoric/documentation/issues/582

**Comment by dckc** [comment]
## hardened js vs. stackblitz

tried dapp-agoric-basics on stackblitz...

```
~/Agoric/dapp-agoric-basics/contract main
❯ yarn test
yarn run v1.22.19
$ ava --verbose

  Uncaught exception in test/test-build-proposal.js

  Error: Accessor not expected at intrinsics.%IteratorPrototype%.constructor
    at async _0x4300f1._evaluate (https://dappagoricbasics-1nwp.w-credentialless-staticblitz.com/blitz.810981ba.js:352:378357)
    at async _0x4300f1._evaluate (https://dappagoricbasics-1nwp.w-credentialless-staticblitz.com/blitz.810981ba.js:352:378357)
    at async _0x4300f1._evaluate (https://dappagoricbasics-1nwp.w-credentialless-staticblitz.com/blitz.810981ba.js:352:378357)
```

cc @kriskowal 

this is after I hacked yarn.lock to avoid:

```
error An unexpected error occurred: "https://codeload.github.com/agoric-labs/esm/tar.gz/3603726ad4636b2f865f463188fcaade6375638e: socket hang up".
```

**Comment by dckc** [comment]
## progress using a3p on gitpod

haven't checked that everything is wired up, but a blockchain and a ui are running...

<img width=400 src="https://github.com/Agoric/agoric-sdk/assets/150986/4612dfa0-b5c1-4169-a0ce-7e32a4a48022">

[8b1ac30](https://github.com/dckc/agoric-gitpod-nix/commit/8b1ac30cdbd9b52e7f82a9956257568ba31b4f86)
https://github.com/dckc/agoric-gitpod-nix

manually cloned dapp-offer-up

**Comment by dckc** [comment]
## websocket?

what code is trying to make a websocket connection?

```
App.tsx:122 DOMException: Failed to construct 'WebSocket': The URL 'ws://https:26657-dckc-agoricgitpodnix-p8ap5wv8qx6.ws-us110.gitpod.io/websocket' is invalid.
  at async connectWallet (https://5173-dckc-agoricgitpodnix-p8ap5wv8qx6.ws-us110.gitpod.io/src/App.tsx?t=1713589839756:65:18)
```

## keplr suggest chain

I wonder how to get this to work.

the `local.agoric.something` hack seems to still be there?

**Comment by dckc** [comment]
https://idx.google.com/dapp-orchestration-basics-5074688
